ceph-releases/luminous/ubuntu/16.04/daemon/zap_device.sh

```diff
@@ -11,70 +101,34 @@ function zap_device {
-        exit 1
-      fi
-    done
+    OSD_DEVICE=$(get_all_ceph_devices)
+  fi
+  # testing all the devices first so we just don't do anything if one device is wrong
+  for device in $(comma_to_space "${OSD_DEVICE}"); do
+    # Check if ${device} is well a block device
+    [[ -b "${device}" ]] || log "Provided device ${device} does not exist or is not a valid block device."
 
-    # look for Ceph encrypted partitions
-    local ceph_dm
-    ceph_dm=$(blkid -t TYPE="crypto_LUKS" "${OSD_DEVICE}"* -o value -s PARTUUID || true)
+    partitions=$(get_child_partitions "${device}")
+    # if the disk passed is a raw device AND the boot system disk
+    [[ $(lsblk --nodeps -no LABEL "${1}") == "boot" ]] && log "Looks like ${device} has a boot partition," &&
+      log "if you want to delete specific partitions point to the partition instead of the raw device" &&
+      log "Do not use your system disk!" &&
+      exit 1
+    ceph_dm=$(get_dmcrypt_uuid_part "${device}")
+    opened_dm=$(get_opened_dmcrypt "${device}")
+    # If dmcrypt partitions detected, loop over all uuid found and check whether they are still opened.
     if [[ -n $ceph_dm ]]; then
-      for dm_uuid in $ceph_dm; do
-        local dm_path="/dev/disk/by-partuuid/$dm_uuid"
-        dmsetup --verbose --force wipe_table "$dm_uuid" || true
-        dmsetup --verbose --force remove "$dm_uuid" || true
-
-        # erase all keyslots (remove encryption key)
-        cryptsetup --verbose --batch-mode erase "$dm_path"
-        local payload_offset
-        payload_offset=$(cryptsetup luksDump "$dm_path" | awk '/Payload offset:/ { print $3 }')
-        local phys_sector_size
-        phys_sector_size=$(blockdev --getpbsz "$dm_path")
-        if ! is_integer "$phys_sector_size"; then
-          # If the sector size isn't a number, let's default to 512
-          phys_sector_size=512
-        fi
-        # remove LUKS header
-        dd if=/dev/zero of="$dm_path" bs="$phys_sector_size" count="$payload_offset" oflag=direct
-      done
+      zap_dmcrypt_device "$ceph_dm" "$opened_dm"
     fi
-
-    for device in $(comma_to_space "${OSD_DEVICE}"); do
-      local raw_device
-      raw_device=$(echo "$device" | grep -oE "${device_match_string}")
-      if echo "$device" | grep -sqE "${device_match_string}"; then
-        log "Zapping the entire device $device"
-        sgdisk --zap-all --clear --mbrtogpt -g -- "$device"
-      else
-        # get the desired partition number(s)
-        local partition_nb
-        partition_nb=$(echo "$device" | grep -oE '[0-9]{1,2}$')
-        log "Zapping partition $device"
-        sgdisk --delete "$partition_nb" "$raw_device"
-      fi
-      log "Executing partprobe on $raw_device"
-      partprobe "$raw_device"
-      udevadm settle
+    log "Zapping the entire device ${device}"
+    for p in $partitions; do
+      dd if=/dev/zero of="${p}" bs=1 count=4096
     done
-  fi
+    sgdisk --zap-all --clear --mbrtogpt -g -- "${device}"
```

Comment:
ceph-disk does a `wipefs` call too which is useful. See http://tracker.ceph.com/issues/6258

Reply:
```
                    'wipefs',
                    '--all',
                    partition,
```